From 5351c71cc75267de36c0485fd679368f311f3d25 Mon Sep 17 00:00:00 2001
From: looooo <sppedflyer@gmail.com>
Date: Sat, 1 Oct 2016 23:36:39 +0200
Subject: [PATCH 1/2] patch

---
 cmake/Modules/FindPythonInterpWithDebug.cmake | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/cmake/Modules/FindPythonInterpWithDebug.cmake b/cmake/Modules/FindPythonInterpWithDebug.cmake
index 7722d95..8407522 100644
--- a/cmake/Modules/FindPythonInterpWithDebug.cmake
+++ b/cmake/Modules/FindPythonInterpWithDebug.cmake
@@ -27,7 +27,7 @@ endif()
 # Detect if the python libs were compiled in debug mode
 execute_process(
     COMMAND ${PYTHON_EXECUTABLE} -c "from distutils import sysconfig; \\
-        print bool(sysconfig.get_config_var('Py_DEBUG'))"
+        print(bool(sysconfig.get_config_var('Py_DEBUG'))")
     OUTPUT_VARIABLE PYTHON_WITH_DEBUG
     OUTPUT_STRIP_TRAILING_WHITESPACE)
 
@@ -36,7 +36,7 @@ execute_process(
         from distutils import sysconfig; \\
         vr = sys.version_info; \\
         suffix = '-dbg' if bool(sysconfig.get_config_var('Py_DEBUG')) else ''; \\
-        print '-python%d.%d%s' % (vr[0], vr[1], suffix)"
+        print('-python%d.%d%s' % (vr[0], vr[1], suffix)")
     OUTPUT_VARIABLE PYTHON_SUFFIX
     OUTPUT_STRIP_TRAILING_WHITESPACE)
 
-- 
2.7.4


From f8359af6a7c4bbff6d16af719d08bde49b9bb272 Mon Sep 17 00:00:00 2001
From: looooo <sppedflyer@gmail.com>
Date: Sat, 1 Oct 2016 23:45:12 +0200
Subject: [PATCH 2/2] patch

---
 cmake/Modules/FindPythonInterpWithDebug.cmake | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/cmake/Modules/FindPythonInterpWithDebug.cmake b/cmake/Modules/FindPythonInterpWithDebug.cmake
index 8407522..88c1e66 100644
--- a/cmake/Modules/FindPythonInterpWithDebug.cmake
+++ b/cmake/Modules/FindPythonInterpWithDebug.cmake
@@ -27,7 +27,7 @@ endif()
 # Detect if the python libs were compiled in debug mode
 execute_process(
     COMMAND ${PYTHON_EXECUTABLE} -c "from distutils import sysconfig; \\
-        print(bool(sysconfig.get_config_var('Py_DEBUG'))")
+        print(bool(sysconfig.get_config_var('Py_DEBUG')))"
     OUTPUT_VARIABLE PYTHON_WITH_DEBUG
     OUTPUT_STRIP_TRAILING_WHITESPACE)
 
@@ -36,7 +36,7 @@ execute_process(
         from distutils import sysconfig; \\
         vr = sys.version_info; \\
         suffix = '-dbg' if bool(sysconfig.get_config_var('Py_DEBUG')) else ''; \\
-        print('-python%d.%d%s' % (vr[0], vr[1], suffix)")
+        print('-python%d.%d%s' % (vr[0], vr[1], suffix))"
     OUTPUT_VARIABLE PYTHON_SUFFIX
     OUTPUT_STRIP_TRAILING_WHITESPACE)
 
-- 
2.7.4

