diff --git a/CMakeLists.txt b/CMakeLists.txt
index 371456874..408678e1c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -43,12 +43,12 @@ else()
 		set(SIMAGE_RUNTIME_LINKING OFF)
 		if(WIN32)
 			if(COIN_LINK_LIBSIMAGE STREQUAL "dynamic")
-				link_libraries(debug simage1d)
-				link_libraries(optimized simage1)
+				link_libraries(debug simaged)
+				link_libraries(optimized simage)
 				add_definitions(-DSIMAGE_DLL)
 			elseif(COIN_LINK_LIBSIMAGE STREQUAL "static")
-				link_libraries(debug simage1sd)
-				link_libraries(optimized simage1s)
+				link_libraries(debug simagesd)
+				link_libraries(optimized simages)
 				add_definitions(-DSIMAGE_NOT_DLL)
 				link_libraries(gdiplus)
 			endif()
