diff --git a/CMakeLists.txt b/CMakeLists.txt
index 076fa021b..002f29568 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -611,7 +611,7 @@ endif(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
         )
 
         # check which modules are available
-        if(UNIX)
+        if(UNIX OR WIN32)
             find_package(VTK COMPONENTS vtkCommonCore REQUIRED NO_MODULE)
             list(APPEND VTK_COMPONENTS vtkIOMPIParallel vtkParallelMPI vtkhdf5)
             foreach(_module ${VTK_COMPONENTS})
diff --git a/src/3rdParty/salomesmesh/CMakeLists.txt b/src/3rdParty/salomesmesh/CMakeLists.txt
index 7818e66..9013cfc 100644
--- a/src/3rdParty/salomesmesh/CMakeLists.txt
+++ b/src/3rdParty/salomesmesh/CMakeLists.txt
@@ -72,6 +72,16 @@ if(MSVC)
             ${OCC_OCAF_LIBRARIES}
         )
     endif(BUILD_FEM_NETGEN)
+
+    #Universal C runtime introduced in VS 2015 (cl version 19)
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "18")
+        list(APPEND SMESH_LIBS
+            debug vcruntimed.lib
+            debug ucrtd.lib
+            optimized vcruntime.lib
+            optimized ucrt.lib
+        )
+    endif()
 else(MSVC)
     if(BUILD_FEM_NETGEN)
         set(SMESH_LIBS
diff --git a/src/Base/CMakeLists.txt b/src/Base/CMakeLists.txt
index 68f8170..89f8d01 100644
--- a/src/Base/CMakeLists.txt
+++ b/src/Base/CMakeLists.txt
@@ -30,6 +30,16 @@ if(MSVC)
         Rpcrt4.lib
         version.lib
     )
+
+    #Universal C runtime introduced in VS 2015 (cl version 19)
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "18")
+        list(APPEND FreeCADBase_LIBS
+            debug vcruntimed.lib
+            debug ucrtd.lib
+            optimized vcruntime.lib
+            optimized ucrt.lib
+        )
+    endif()
 elseif(MINGW)
     set(FreeCADBase_LIBS
         ${PYTHON_LIBRARIES}
diff --git a/src/Mod/Path/libarea/CMakeLists.txt b/src/Mod/Path/libarea/CMakeLists.txt
index 6e489b6..a645ca3 100644
--- a/src/Mod/Path/libarea/CMakeLists.txt
+++ b/src/Mod/Path/libarea/CMakeLists.txt
@@ -84,6 +84,17 @@ if(MSVC)
         optimized MSVCRT.LIB
         optimized MSVCPRT.LIB
     )
+
+    #Universal C runtime introduced in VS 2015 (cl version 19)
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "18")
+        list(APPEND area_native_LIBS
+            debug vcruntimed.lib
+            debug ucrtd.lib
+            optimized vcruntime.lib
+            optimized ucrt.lib
+        )
+    endif()
+
     set(area_LIBS
         ${Boost_LIBRARIES}
         ${PYTHON_LIBRARIES}
