diff --git a/cMake/FindCoin3D.cmake b/cMake/FindCoin3D.cmake
index 2680b694e..8327f3944 100644
--- a/cMake/FindCoin3D.cmake
+++ b/cMake/FindCoin3D.cmake
@@ -12,12 +12,14 @@ IF (WIN32)
   IF (CYGWIN)
 
     FIND_PATH(COIN3D_INCLUDE_DIRS Inventor/So.h
+      ${CMAKE_INCLUDE_PATH}
       /usr/include
       /usr/local/include
       /usr/include/coin
     )
 
     FIND_LIBRARY(COIN3D_LIBRARIES Coin
+      ${CMAKE_LIBRARY_PATH}
       /usr/lib
       /usr/local/lib
     )
@@ -25,6 +27,7 @@ IF (WIN32)
   ELSE (CYGWIN)
 
     FIND_PATH(COIN3D_INCLUDE_DIRS Inventor/So.h
+      ${CMAKE_INCLUDE_PATH}
       "[HKEY_LOCAL_MACHINE\\SOFTWARE\\SIM\\Coin3D\\2;Installation Path]/include"
     )
 
@@ -32,7 +35,12 @@ IF (WIN32)
       "[HKEY_LOCAL_MACHINE\\SOFTWARE\\SIM\\Coin3D\\2;Installation Path]/lib"
     )
 
-    FIND_LIBRARY(COIN3D_LIBRARY_RELEASE coin2
+    FIND_LIBRARY(COIN3D_LIBRARY_RELEASE 
+    NAMES
+      coin2
+      Coin4
+    PATHS
+      ${CMAKE_LIBRARY_PATH}
       "[HKEY_LOCAL_MACHINE\\SOFTWARE\\SIM\\Coin3D\\2;Installation Path]/lib"
     )
 
