diff --git a/src/CXX/Python3/IndirectPythonInterface.hxx b/src/CXX/Python3/IndirectPythonInterface.hxx
index 5ffaf43..a4c0cf4 100644
--- a/src/CXX/Python3/IndirectPythonInterface.hxx
+++ b/src/CXX/Python3/IndirectPythonInterface.hxx
@@ -175,8 +175,8 @@ PYCXX_EXPORT int &_Py_TabcheckFlag();
 PYCXX_EXPORT int &_Py_VerboseFlag();
 PYCXX_EXPORT int &_Py_UnicodeFlag();
 
-void _XINCREF( PyObject *op );
-void _XDECREF( PyObject *op );
+PYCXX_EXPORT void _XINCREF( PyObject *op );
+PYCXX_EXPORT void _XDECREF( PyObject *op );
 
 PYCXX_EXPORT char *__Py_PackageContext();
 };
diff --git a/src/CXX/Python3/Objects.hxx b/src/CXX/Python3/Objects.hxx
index 8f62eac..cf26be9 100644
--- a/src/CXX/Python3/Objects.hxx
+++ b/src/CXX/Python3/Objects.hxx
@@ -429,12 +429,12 @@ namespace Py
     };
 
     //------------------------------------------------------------
-    bool operator==( const Object &o1, const Object &o2 );
-    bool operator!=( const Object &o1, const Object &o2 );
-    bool operator>=( const Object &o1, const Object &o2 );
-    bool operator<=( const Object &o1, const Object &o2 );
-    bool operator<( const Object &o1, const Object &o2 );
-    bool operator>( const Object &o1, const Object &o2 );
+    PYCXX_EXPORT bool operator==( const Object &o1, const Object &o2 );
+    PYCXX_EXPORT bool operator!=( const Object &o1, const Object &o2 );
+    PYCXX_EXPORT bool operator>=( const Object &o1, const Object &o2 );
+    PYCXX_EXPORT bool operator<=( const Object &o1, const Object &o2 );
+    PYCXX_EXPORT bool operator<( const Object &o1, const Object &o2 );
+    PYCXX_EXPORT bool operator>( const Object &o1, const Object &o2 );
 
     //------------------------------------------------------------
 
