From 4ce209beeca14aea3b263d36c1ceca1fcb7d7b02 Mon Sep 17 00:00:00 2001
From: looooo <sppedflyer@gmail.com>
Date: Tue, 16 Aug 2016 15:04:06 +0200
Subject: [PATCH 1/7] occ cmake patch

---
 cMake/FindOpenCasCade.cmake | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/cMake/FindOpenCasCade.cmake b/cMake/FindOpenCasCade.cmake
index fbcc262..d7bd7c4 100644
--- a/cMake/FindOpenCasCade.cmake
+++ b/cMake/FindOpenCasCade.cmake
@@ -6,6 +6,13 @@
 # OCC_LIBRARY_DIR    - where the OCC library directory can be found
 # OCC_LIBRARIES      - Link this to use OCC
 # OCC_OCAF_LIBRARIES - Link this to use OCC OCAF framework
+if(TRUE)
+	set(OCC_FOUND TRUE)
+	set(OCC_INCLUDE_DIR ${CMAKE_INSTALL_PREFIX}/include/opencascade)
+	set(OCC_LIBRARY_DIR ${CMAKE_INSTALL_PREFIX}/lib)
+	set(OCC_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
+	set(OCC_OCAF_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
+else(True)
 
 # First try to find OpenCASCADE Community Edition
 if(NOT DEFINED OCE_DIR)
@@ -142,3 +149,5 @@ if(OCC_FOUND)
 else(OCC_FOUND)
   #message(SEND_ERROR "Neither OpenCASCADE Community Edition nor OpenCasCade were found: will not build CAD modules!")
 endif(OCC_FOUND)
+
+endif(True)
\ No newline at end of file
-- 
2.7.4


From d3781113be90289bfa2cf5fe7ee6cf9d61f2b5a4 Mon Sep 17 00:00:00 2001
From: looooo <sppedflyer@gmail.com>
Date: Tue, 16 Aug 2016 15:09:58 +0200
Subject: [PATCH 2/7] occ_cmake

---
 cMake/FindOpenCasCade.cmake | 275 ++++++++++++++++++++++----------------------
 1 file changed, 137 insertions(+), 138 deletions(-)

diff --git a/cMake/FindOpenCasCade.cmake b/cMake/FindOpenCasCade.cmake
index d7bd7c4..7b763c2 100644
--- a/cMake/FindOpenCasCade.cmake
+++ b/cMake/FindOpenCasCade.cmake
@@ -6,148 +6,147 @@
 # OCC_LIBRARY_DIR    - where the OCC library directory can be found
 # OCC_LIBRARIES      - Link this to use OCC
 # OCC_OCAF_LIBRARIES - Link this to use OCC OCAF framework
-if(TRUE)
-	set(OCC_FOUND TRUE)
-	set(OCC_INCLUDE_DIR ${CMAKE_INSTALL_PREFIX}/include/opencascade)
-	set(OCC_LIBRARY_DIR ${CMAKE_INSTALL_PREFIX}/lib)
-	set(OCC_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
-	set(OCC_OCAF_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
-else(True)
+set(OCC_FOUND TRUE)
+set(OCC_INCLUDE_DIR ${CMAKE_INSTALL_PREFIX}/include/opencascade)
+set(OCC_LIBRARY_DIR ${CMAKE_INSTALL_PREFIX}/lib)
+set(OCC_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
+set(OCC_OCAF_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
 
-# First try to find OpenCASCADE Community Edition
-if(NOT DEFINED OCE_DIR)
-  # Check for OSX needs to come first because UNIX evaluates to true on OSX
-  if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
-    if(DEFINED MACPORTS_PREFIX)
-      find_package(OCE QUIET HINTS ${MACPORTS_PREFIX}/Library/Frameworks)
-    elseif(DEFINED HOMEBREW_PREFIX)
-      find_package(OCE QUIET HINTS ${HOMEBREW_PREFIX}/Cellar/oce/*)
-    endif()
-  elseif(UNIX)
-    set(OCE_DIR "/usr/local/share/cmake/")
-  elseif(WIN32)
-    set(OCE_DIR "c:/OCE-0.4.0/share/cmake")
-  endif()
-endif()
 
-find_package(OCE QUIET)
-if(OCE_FOUND)
-  message(STATUS "-- OpenCASCADE Community Edition has been found.")
-  # Disable this define. For more details see bug #0001872
-  #add_definitions (-DHAVE_CONFIG_H)
-  set(OCC_INCLUDE_DIR ${OCE_INCLUDE_DIRS})
-  #set(OCC_LIBRARY_DIR ${OCE_LIBRARY_DIR})
-else(OCE_FOUND) #look for OpenCASCADE
-  if(WIN32)
-    if(CYGWIN OR MINGW)
-    FIND_PATH(OCC_INCLUDE_DIR Standard_Version.hxx
-        /usr/include/opencascade
-        /usr/local/include/opencascade
-        /opt/opencascade/include
-        /opt/opencascade/inc
-      )
-      FIND_LIBRARY(OCC_LIBRARY TKernel
-        /usr/lib
-        /usr/local/lib
-        /opt/opencascade/lib
-      )
-    else(CYGWIN OR MINGW)
-    FIND_PATH(OCC_INCLUDE_DIR Standard_Version.hxx
-        "[HKEY_LOCAL_MACHINE\\SOFTWARE\\SIM\\OCC\\2;Installation Path]/include"
-      )
-      FIND_LIBRARY(OCC_LIBRARY TKernel
-        "[HKEY_LOCAL_MACHINE\\SOFTWARE\\SIM\\OCC\\2;Installation Path]/lib"
-      )
-    endif(CYGWIN OR MINGW)
-  else(WIN32)
-    FIND_PATH(OCC_INCLUDE_DIR Standard_Version.hxx
-      /usr/include/opencascade
-      /usr/local/include/opencascade
-      /opt/opencascade/include
-      /opt/opencascade/inc
-    )
-    FIND_LIBRARY(OCC_LIBRARY TKernel
-      /usr/lib
-      /usr/local/lib
-      /opt/opencascade/lib
-    )
-  endif(WIN32)
-  if(OCC_LIBRARY)
-    GET_FILENAME_COMPONENT(OCC_LIBRARY_DIR ${OCC_LIBRARY} PATH)
-    IF(NOT OCC_INCLUDE_DIR)
-      FIND_PATH(OCC_INCLUDE_DIR Standard_Version.hxx
-        ${OCC_LIBRARY_DIR}/../inc
-      )
-    ENDIF()
-  endif(OCC_LIBRARY)
-endif(OCE_FOUND)
+# # First try to find OpenCASCADE Community Edition
+# if(NOT DEFINED OCE_DIR)
+#   # Check for OSX needs to come first because UNIX evaluates to true on OSX
+#   if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+#     if(DEFINED MACPORTS_PREFIX)
+#       find_package(OCE QUIET HINTS ${MACPORTS_PREFIX}/Library/Frameworks)
+#     elseif(DEFINED HOMEBREW_PREFIX)
+#       find_package(OCE QUIET HINTS ${HOMEBREW_PREFIX}/Cellar/oce/*)
+#     endif()
+#   elseif(UNIX)
+#     set(OCE_DIR "/usr/local/share/cmake/")
+#   elseif(WIN32)
+#     set(OCE_DIR "c:/OCE-0.4.0/share/cmake")
+#   endif()
+# endif()
 
-if(OCC_INCLUDE_DIR)
-  file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MAJOR
-    REGEX "#define OCC_VERSION_MAJOR.*"
-  )
-  string(REGEX MATCH "[0-9]+" OCC_MAJOR ${OCC_MAJOR})
-  file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MINOR
-    REGEX "#define OCC_VERSION_MINOR.*"
-  )
-  string(REGEX MATCH "[0-9]+" OCC_MINOR ${OCC_MINOR})
-  file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MAINT
-    REGEX "#define OCC_VERSION_MAINTENANCE.*"
-  )
-  string(REGEX MATCH "[0-9]+" OCC_MAINT ${OCC_MAINT})
+# find_package(OCE QUIET)
+# if(OCE_FOUND)
+#   message(STATUS "-- OpenCASCADE Community Edition has been found.")
+#   # Disable this define. For more details see bug #0001872
+#   #add_definitions (-DHAVE_CONFIG_H)
+#   set(OCC_INCLUDE_DIR ${OCE_INCLUDE_DIRS})
+#   #set(OCC_LIBRARY_DIR ${OCE_LIBRARY_DIR})
+# else(OCE_FOUND) #look for OpenCASCADE
+#   if(WIN32)
+#     if(CYGWIN OR MINGW)
+#     FIND_PATH(OCC_INCLUDE_DIR Standard_Version.hxx
+#         /usr/include/opencascade
+#         /usr/local/include/opencascade
+#         /opt/opencascade/include
+#         /opt/opencascade/inc
+#       )
+#       FIND_LIBRARY(OCC_LIBRARY TKernel
+#         /usr/lib
+#         /usr/local/lib
+#         /opt/opencascade/lib
+#       )
+#     else(CYGWIN OR MINGW)
+#     FIND_PATH(OCC_INCLUDE_DIR Standard_Version.hxx
+#         "[HKEY_LOCAL_MACHINE\\SOFTWARE\\SIM\\OCC\\2;Installation Path]/include"
+#       )
+#       FIND_LIBRARY(OCC_LIBRARY TKernel
+#         "[HKEY_LOCAL_MACHINE\\SOFTWARE\\SIM\\OCC\\2;Installation Path]/lib"
+#       )
+#     endif(CYGWIN OR MINGW)
+#   else(WIN32)
+#     FIND_PATH(OCC_INCLUDE_DIR Standard_Version.hxx
+#       /usr/include/opencascade
+#       /usr/local/include/opencascade
+#       /opt/opencascade/include
+#       /opt/opencascade/inc
+#     )
+#     FIND_LIBRARY(OCC_LIBRARY TKernel
+#       /usr/lib
+#       /usr/local/lib
+#       /opt/opencascade/lib
+#     )
+#   endif(WIN32)
+#   if(OCC_LIBRARY)
+#     GET_FILENAME_COMPONENT(OCC_LIBRARY_DIR ${OCC_LIBRARY} PATH)
+#     IF(NOT OCC_INCLUDE_DIR)
+#       FIND_PATH(OCC_INCLUDE_DIR Standard_Version.hxx
+#         ${OCC_LIBRARY_DIR}/../inc
+#       )
+#     ENDIF()
+#   endif(OCC_LIBRARY)
+# endif(OCE_FOUND)
 
-  set(OCC_VERSION_STRING "${OCC_MAJOR}.${OCC_MINOR}.${OCC_MAINT}")
-endif(OCC_INCLUDE_DIR)
+# if(OCC_INCLUDE_DIR)
+#   file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MAJOR
+#     REGEX "#define OCC_VERSION_MAJOR.*"
+#   )
+#   string(REGEX MATCH "[0-9]+" OCC_MAJOR ${OCC_MAJOR})
+#   file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MINOR
+#     REGEX "#define OCC_VERSION_MINOR.*"
+#   )
+#   string(REGEX MATCH "[0-9]+" OCC_MINOR ${OCC_MINOR})
+#   file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MAINT
+#     REGEX "#define OCC_VERSION_MAINTENANCE.*"
+#   )
+#   string(REGEX MATCH "[0-9]+" OCC_MAINT ${OCC_MAINT})
 
-# handle the QUIETLY and REQUIRED arguments and set OCC_FOUND to TRUE if
-# all listed variables are TRUE
-include(FindPackageHandleStandardArgs)
-FIND_PACKAGE_HANDLE_STANDARD_ARGS(OCC REQUIRED_VARS OCC_INCLUDE_DIR VERSION_VAR OCC_VERSION_STRING)
+#   set(OCC_VERSION_STRING "${OCC_MAJOR}.${OCC_MINOR}.${OCC_MAINT}")
+# endif(OCC_INCLUDE_DIR)
 
-if(OCC_FOUND)
-  set(OCC_LIBRARIES
-    TKFillet
-    TKMesh
-    TKernel
-    TKG2d
-    TKG3d
-    TKMath
-    TKIGES
-    TKSTL
-    TKShHealing
-    TKXSBase
-    TKBin
-    TKBool
-    TKBO
-    TKCDF
-    TKBRep
-    TKTopAlgo
-    TKGeomAlgo
-    TKGeomBase
-    TKOffset
-    TKPrim
-    TKSTEP
-    TKSTEPBase
-    TKSTEPAttr
-    TKHLR
-    TKFeat
-  )
-  set(OCC_OCAF_LIBRARIES
-    TKCAF
-    TKXCAF
-    TKLCAF
-    TKXDESTEP
-    TKXDEIGES
-    TKMeshVS
-  )
-  if(OCC_VERSION_STRING VERSION_LESS 6.7.3)
-    list(APPEND OCC_OCAF_LIBRARIES TKAdvTools)
-  endif(OCC_VERSION_STRING VERSION_LESS 6.7.3)
-  message(STATUS "-- Found OCE/OpenCASCADE version: ${OCC_VERSION_STRING}")
-  message(STATUS "-- OCE/OpenCASCADE include directory: ${OCC_INCLUDE_DIR}")
-  message(STATUS "-- OCE/OpenCASCADE shared libraries directory: ${OCC_LIBRARY_DIR}")
-else(OCC_FOUND)
-  #message(SEND_ERROR "Neither OpenCASCADE Community Edition nor OpenCasCade were found: will not build CAD modules!")
-endif(OCC_FOUND)
+# # handle the QUIETLY and REQUIRED arguments and set OCC_FOUND to TRUE if
+# # all listed variables are TRUE
+# include(FindPackageHandleStandardArgs)
+# FIND_PACKAGE_HANDLE_STANDARD_ARGS(OCC REQUIRED_VARS OCC_INCLUDE_DIR VERSION_VAR OCC_VERSION_STRING)
 
-endif(True)
\ No newline at end of file
+# if(OCC_FOUND)
+#   set(OCC_LIBRARIES
+#     TKFillet
+#     TKMesh
+#     TKernel
+#     TKG2d
+#     TKG3d
+#     TKMath
+#     TKIGES
+#     TKSTL
+#     TKShHealing
+#     TKXSBase
+#     TKBin
+#     TKBool
+#     TKBO
+#     TKCDF
+#     TKBRep
+#     TKTopAlgo
+#     TKGeomAlgo
+#     TKGeomBase
+#     TKOffset
+#     TKPrim
+#     TKSTEP
+#     TKSTEPBase
+#     TKSTEPAttr
+#     TKHLR
+#     TKFeat
+#   )
+#   set(OCC_OCAF_LIBRARIES
+#     TKCAF
+#     TKXCAF
+#     TKLCAF
+#     TKXDESTEP
+#     TKXDEIGES
+#     TKMeshVS
+#   )
+#   if(OCC_VERSION_STRING VERSION_LESS 6.7.3)
+#     list(APPEND OCC_OCAF_LIBRARIES TKAdvTools)
+#   endif(OCC_VERSION_STRING VERSION_LESS 6.7.3)
+#   message(STATUS "-- Found OCE/OpenCASCADE version: ${OCC_VERSION_STRING}")
+#   message(STATUS "-- OCE/OpenCASCADE include directory: ${OCC_INCLUDE_DIR}")
+#   message(STATUS "-- OCE/OpenCASCADE shared libraries directory: ${OCC_LIBRARY_DIR}")
+# else(OCC_FOUND)
+#   #message(SEND_ERROR "Neither OpenCASCADE Community Edition nor OpenCasCade were found: will not build CAD modules!")
+# endif(OCC_FOUND)
+
+# endif(True)
-- 
2.7.4


From 29c30718acae1d125ef86c7aa101621328fd5ed3 Mon Sep 17 00:00:00 2001
From: looooo <sppedflyer@gmail.com>
Date: Tue, 16 Aug 2016 15:40:11 +0200
Subject: [PATCH 3/7] patch

---
 cMake/FindOpenCasCade.cmake | 92 ++++++++++++++++++++++-----------------------
 1 file changed, 45 insertions(+), 47 deletions(-)

diff --git a/cMake/FindOpenCasCade.cmake b/cMake/FindOpenCasCade.cmake
index 7b763c2..2bb42af 100644
--- a/cMake/FindOpenCasCade.cmake
+++ b/cMake/FindOpenCasCade.cmake
@@ -103,50 +103,48 @@ set(OCC_OCAF_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
 # include(FindPackageHandleStandardArgs)
 # FIND_PACKAGE_HANDLE_STANDARD_ARGS(OCC REQUIRED_VARS OCC_INCLUDE_DIR VERSION_VAR OCC_VERSION_STRING)
 
-# if(OCC_FOUND)
-#   set(OCC_LIBRARIES
-#     TKFillet
-#     TKMesh
-#     TKernel
-#     TKG2d
-#     TKG3d
-#     TKMath
-#     TKIGES
-#     TKSTL
-#     TKShHealing
-#     TKXSBase
-#     TKBin
-#     TKBool
-#     TKBO
-#     TKCDF
-#     TKBRep
-#     TKTopAlgo
-#     TKGeomAlgo
-#     TKGeomBase
-#     TKOffset
-#     TKPrim
-#     TKSTEP
-#     TKSTEPBase
-#     TKSTEPAttr
-#     TKHLR
-#     TKFeat
-#   )
-#   set(OCC_OCAF_LIBRARIES
-#     TKCAF
-#     TKXCAF
-#     TKLCAF
-#     TKXDESTEP
-#     TKXDEIGES
-#     TKMeshVS
-#   )
-#   if(OCC_VERSION_STRING VERSION_LESS 6.7.3)
-#     list(APPEND OCC_OCAF_LIBRARIES TKAdvTools)
-#   endif(OCC_VERSION_STRING VERSION_LESS 6.7.3)
-#   message(STATUS "-- Found OCE/OpenCASCADE version: ${OCC_VERSION_STRING}")
-#   message(STATUS "-- OCE/OpenCASCADE include directory: ${OCC_INCLUDE_DIR}")
-#   message(STATUS "-- OCE/OpenCASCADE shared libraries directory: ${OCC_LIBRARY_DIR}")
-# else(OCC_FOUND)
-#   #message(SEND_ERROR "Neither OpenCASCADE Community Edition nor OpenCasCade were found: will not build CAD modules!")
-# endif(OCC_FOUND)
-
-# endif(True)
+if(OCC_FOUND)
+  set(OCC_LIBRARIES
+    TKFillet
+    TKMesh
+    TKernel
+    TKG2d
+    TKG3d
+    TKMath
+    TKIGES
+    TKSTL
+    TKShHealing
+    TKXSBase
+    TKBin
+    TKBool
+    TKBO
+    TKCDF
+    TKBRep
+    TKTopAlgo
+    TKGeomAlgo
+    TKGeomBase
+    TKOffset
+    TKPrim
+    TKSTEP
+    TKSTEPBase
+    TKSTEPAttr
+    TKHLR
+    TKFeat
+  )
+  set(OCC_OCAF_LIBRARIES
+    TKCAF
+    TKXCAF
+    TKLCAF
+    TKXDESTEP
+    TKXDEIGES
+    TKMeshVS
+  )
+  if(OCC_VERSION_STRING VERSION_LESS 6.7.3)
+    list(APPEND OCC_OCAF_LIBRARIES TKAdvTools)
+  endif(OCC_VERSION_STRING VERSION_LESS 6.7.3)
+  message(STATUS "-- Found OCE/OpenCASCADE version: ${OCC_VERSION_STRING}")
+  message(STATUS "-- OCE/OpenCASCADE include directory: ${OCC_INCLUDE_DIR}")
+  message(STATUS "-- OCE/OpenCASCADE shared libraries directory: ${OCC_LIBRARY_DIR}")
+else(OCC_FOUND)
+  #message(SEND_ERROR "Neither OpenCASCADE Community Edition nor OpenCasCade were found: will not build CAD modules!")
+endif(OCC_FOUND)
-- 
2.7.4


From 314c2bcb8d664377c0f3dcfc72b9421b4ad610ce Mon Sep 17 00:00:00 2001
From: looooo <sppedflyer@gmail.com>
Date: Tue, 16 Aug 2016 15:42:06 +0200
Subject: [PATCH 4/7] patch

---
 cMake/FindOpenCasCade.cmake | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/cMake/FindOpenCasCade.cmake b/cMake/FindOpenCasCade.cmake
index 2bb42af..e353b1b 100644
--- a/cMake/FindOpenCasCade.cmake
+++ b/cMake/FindOpenCasCade.cmake
@@ -7,10 +7,10 @@
 # OCC_LIBRARIES      - Link this to use OCC
 # OCC_OCAF_LIBRARIES - Link this to use OCC OCAF framework
 set(OCC_FOUND TRUE)
-set(OCC_INCLUDE_DIR ${CMAKE_INSTALL_PREFIX}/include/opencascade)
-set(OCC_LIBRARY_DIR ${CMAKE_INSTALL_PREFIX}/lib)
-set(OCC_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
-set(OCC_OCAF_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
+set(OCC_INCLUDE_DIR ${CMAKE_INSTALL_PREFIX}/../include/opencascade)
+set(OCC_LIBRARY_DIR ${CMAKE_INSTALL_PREFIX}/../lib)
+set(OCC_LIBRARIES ${CMAKE_INSTALL_PREFIX}/../lib)
+set(OCC_OCAF_LIBRARIES ${CMAKE_INSTALL_PREFIX}/../lib)
 
 
 # # First try to find OpenCASCADE Community Edition
-- 
2.7.4


From 587a3700c5daa9bc35e0bbcda8ac469b0e8c0f41 Mon Sep 17 00:00:00 2001
From: looooo <sppedflyer@gmail.com>
Date: Tue, 16 Aug 2016 15:47:24 +0200
Subject: [PATCH 5/7] patch

---
 cMake/FindOpenCasCade.cmake | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/cMake/FindOpenCasCade.cmake b/cMake/FindOpenCasCade.cmake
index e353b1b..2bb42af 100644
--- a/cMake/FindOpenCasCade.cmake
+++ b/cMake/FindOpenCasCade.cmake
@@ -7,10 +7,10 @@
 # OCC_LIBRARIES      - Link this to use OCC
 # OCC_OCAF_LIBRARIES - Link this to use OCC OCAF framework
 set(OCC_FOUND TRUE)
-set(OCC_INCLUDE_DIR ${CMAKE_INSTALL_PREFIX}/../include/opencascade)
-set(OCC_LIBRARY_DIR ${CMAKE_INSTALL_PREFIX}/../lib)
-set(OCC_LIBRARIES ${CMAKE_INSTALL_PREFIX}/../lib)
-set(OCC_OCAF_LIBRARIES ${CMAKE_INSTALL_PREFIX}/../lib)
+set(OCC_INCLUDE_DIR ${CMAKE_INSTALL_PREFIX}/include/opencascade)
+set(OCC_LIBRARY_DIR ${CMAKE_INSTALL_PREFIX}/lib)
+set(OCC_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
+set(OCC_OCAF_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
 
 
 # # First try to find OpenCASCADE Community Edition
-- 
2.7.4


From 80ae0b9095a00d2098ae8f7e6938d4c7602afc50 Mon Sep 17 00:00:00 2001
From: looooo <sppedflyer@gmail.com>
Date: Tue, 16 Aug 2016 15:57:19 +0200
Subject: [PATCH 6/7] patch

---
 cMake/FindOpenCasCade.cmake | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/cMake/FindOpenCasCade.cmake b/cMake/FindOpenCasCade.cmake
index 2bb42af..2daf0c2 100644
--- a/cMake/FindOpenCasCade.cmake
+++ b/cMake/FindOpenCasCade.cmake
@@ -139,9 +139,9 @@ if(OCC_FOUND)
     TKXDEIGES
     TKMeshVS
   )
-  if(OCC_VERSION_STRING VERSION_LESS 6.7.3)
-    list(APPEND OCC_OCAF_LIBRARIES TKAdvTools)
-  endif(OCC_VERSION_STRING VERSION_LESS 6.7.3)
+  # if(OCC_VERSION_STRING VERSION_LESS 6.7.3)
+  #   list(APPEND OCC_OCAF_LIBRARIES TKAdvTools)
+  # endif(OCC_VERSION_STRING VERSION_LESS 6.7.3)
   message(STATUS "-- Found OCE/OpenCASCADE version: ${OCC_VERSION_STRING}")
   message(STATUS "-- OCE/OpenCASCADE include directory: ${OCC_INCLUDE_DIR}")
   message(STATUS "-- OCE/OpenCASCADE shared libraries directory: ${OCC_LIBRARY_DIR}")
-- 
2.7.4


From 4d30c64b8751b2d89126eb88dc6efea757912cfd Mon Sep 17 00:00:00 2001
From: looooo <sppedflyer@gmail.com>
Date: Tue, 16 Aug 2016 15:59:54 +0200
Subject: [PATCH 7/7] patch

---
 cMake/FindOpenCasCade.cmake | 44 ++++++++++++++++++++++----------------------
 1 file changed, 22 insertions(+), 22 deletions(-)

diff --git a/cMake/FindOpenCasCade.cmake b/cMake/FindOpenCasCade.cmake
index 2daf0c2..ff46bd5 100644
--- a/cMake/FindOpenCasCade.cmake
+++ b/cMake/FindOpenCasCade.cmake
@@ -81,27 +81,27 @@ set(OCC_OCAF_LIBRARIES ${CMAKE_INSTALL_PREFIX}/lib)
 #   endif(OCC_LIBRARY)
 # endif(OCE_FOUND)
 
-# if(OCC_INCLUDE_DIR)
-#   file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MAJOR
-#     REGEX "#define OCC_VERSION_MAJOR.*"
-#   )
-#   string(REGEX MATCH "[0-9]+" OCC_MAJOR ${OCC_MAJOR})
-#   file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MINOR
-#     REGEX "#define OCC_VERSION_MINOR.*"
-#   )
-#   string(REGEX MATCH "[0-9]+" OCC_MINOR ${OCC_MINOR})
-#   file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MAINT
-#     REGEX "#define OCC_VERSION_MAINTENANCE.*"
-#   )
-#   string(REGEX MATCH "[0-9]+" OCC_MAINT ${OCC_MAINT})
+if(OCC_INCLUDE_DIR)
+  file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MAJOR
+    REGEX "#define OCC_VERSION_MAJOR.*"
+  )
+  string(REGEX MATCH "[0-9]+" OCC_MAJOR ${OCC_MAJOR})
+  file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MINOR
+    REGEX "#define OCC_VERSION_MINOR.*"
+  )
+  string(REGEX MATCH "[0-9]+" OCC_MINOR ${OCC_MINOR})
+  file(STRINGS ${OCC_INCLUDE_DIR}/Standard_Version.hxx OCC_MAINT
+    REGEX "#define OCC_VERSION_MAINTENANCE.*"
+  )
+  string(REGEX MATCH "[0-9]+" OCC_MAINT ${OCC_MAINT})
 
-#   set(OCC_VERSION_STRING "${OCC_MAJOR}.${OCC_MINOR}.${OCC_MAINT}")
-# endif(OCC_INCLUDE_DIR)
+  set(OCC_VERSION_STRING "${OCC_MAJOR}.${OCC_MINOR}.${OCC_MAINT}")
+endif(OCC_INCLUDE_DIR)
 
-# # handle the QUIETLY and REQUIRED arguments and set OCC_FOUND to TRUE if
-# # all listed variables are TRUE
-# include(FindPackageHandleStandardArgs)
-# FIND_PACKAGE_HANDLE_STANDARD_ARGS(OCC REQUIRED_VARS OCC_INCLUDE_DIR VERSION_VAR OCC_VERSION_STRING)
+# handle the QUIETLY and REQUIRED arguments and set OCC_FOUND to TRUE if
+# all listed variables are TRUE
+include(FindPackageHandleStandardArgs)
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(OCC REQUIRED_VARS OCC_INCLUDE_DIR VERSION_VAR OCC_VERSION_STRING)
 
 if(OCC_FOUND)
   set(OCC_LIBRARIES
@@ -139,9 +139,9 @@ if(OCC_FOUND)
     TKXDEIGES
     TKMeshVS
   )
-  # if(OCC_VERSION_STRING VERSION_LESS 6.7.3)
-  #   list(APPEND OCC_OCAF_LIBRARIES TKAdvTools)
-  # endif(OCC_VERSION_STRING VERSION_LESS 6.7.3)
+  if(OCC_VERSION_STRING VERSION_LESS 6.7.3)
+    list(APPEND OCC_OCAF_LIBRARIES TKAdvTools)
+  endif(OCC_VERSION_STRING VERSION_LESS 6.7.3)
   message(STATUS "-- Found OCE/OpenCASCADE version: ${OCC_VERSION_STRING}")
   message(STATUS "-- OCE/OpenCASCADE include directory: ${OCC_INCLUDE_DIR}")
   message(STATUS "-- OCE/OpenCASCADE shared libraries directory: ${OCC_LIBRARY_DIR}")
-- 
2.7.4

