diff --git a/CMakeLists.txt b/CMakeLists.txt
index e71caca77..6d49222f4 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -910,14 +910,14 @@ endif(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
 
         find_package(PySide2Tools REQUIRED) # PySide2 utilities (pyside2-uic & pyside2-rcc)
 
-        file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/bin/PySide)
-        file(WRITE ${CMAKE_BINARY_DIR}/bin/PySide/__init__.py "# PySide wrapper\n")
-        file(WRITE ${CMAKE_BINARY_DIR}/bin/PySide/QtCore.py "from PySide2.QtCore import *\n\n"
+        file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/Mod/PySide)
+        file(WRITE ${CMAKE_BINARY_DIR}/Mod/PySide/__init__.py "# PySide wrapper\n")
+        file(WRITE ${CMAKE_BINARY_DIR}/Mod/PySide/QtCore.py "from PySide2.QtCore import *\n\n"
                                                             "#QCoreApplication.CodecForTr=0\n"
                                                             "#QCoreApplication.UnicodeUTF8=1\n")
-        file(WRITE ${CMAKE_BINARY_DIR}/bin/PySide/QtGui.py  "from PySide2.QtGui import *\n"
+        file(WRITE ${CMAKE_BINARY_DIR}/Mod/PySide/QtGui.py  "from PySide2.QtGui import *\n"
                                                             "from PySide2.QtWidgets import *\n")
-        file(WRITE ${CMAKE_BINARY_DIR}/bin/PySide/QtSvg.py  "from PySide2.QtSvg import *\n")
+        file(WRITE ${CMAKE_BINARY_DIR}/Mod/PySide/QtSvg.py  "from PySide2.QtSvg import *\n")
 
 	if(APPLE)
 	    INSTALL(
@@ -929,9 +929,9 @@ endif(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
 	else()
             INSTALL(
 	        DIRECTORY	
-                    ${CMAKE_BINARY_DIR}/bin/PySide
+                    ${CMAKE_BINARY_DIR}/Mod/PySide
                 DESTINATION
-                    bin
+                    Mod
             )
 	endif()
     else()
