diff --git a/CMakeLists.txt b/CMakeLists.txt
index 076fa021b..002f29568 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -611,7 +611,7 @@ endif(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
         )
 
         # check which modules are available
-        if(UNIX)
+        if(UNIX OR WIN32)
             find_package(VTK COMPONENTS vtkCommonCore REQUIRED NO_MODULE)
             list(APPEND VTK_COMPONENTS vtkIOMPIParallel vtkParallelMPI vtkhdf5)
             foreach(_module ${VTK_COMPONENTS})
diff --git a/cMake/FindXercesC.cmake b/cMake/FindXercesC.cmake
index f714b5bc9..a4181c41d 100644
--- a/cMake/FindXercesC.cmake
+++ b/cMake/FindXercesC.cmake
@@ -9,6 +9,7 @@
 # XercesC_FOUND, If false, don't try to use pwlib.
 
 FIND_PATH(XercesC_INCLUDE_DIRS xercesc/dom/DOM.hpp
+  ${CMAKE_INCLUDE_PATH}
   "[HKEY_CURRENT_USER\\software\\xerces-c\\src]"
   "[HKEY_CURRENT_USER\\xerces-c\\src]"
   $ENV{XERCESCROOT}/src/
@@ -19,7 +20,9 @@ FIND_PATH(XercesC_INCLUDE_DIRS xercesc/dom/DOM.hpp
 FIND_LIBRARY(XercesC_LIBRARIES
   NAMES
     xerces-c
+    xerces-c_3
   PATHS
+    ${CMAKE_LIBRARY_PATH}
     "[HKEY_CURRENT_USER\\software\\xerces-c\\lib]"
     "[HKEY_CURRENT_USER\\xerces-c\\lib]"
     $ENV{XERCESCROOT}/${LIB_DESTINATION}
