diff --git a/MANIFEST.in b/MANIFEST.in
new file mode 100644
index 0000000..b5bf95b
--- /dev/null
+++ b/MANIFEST.in
@@ -0,0 +1 @@
+recursive-include freecad/asm3/Gui *
diff --git a/freecad/asm3/__init__.py b/freecad/asm3/__init__.py
index e69de29..17c1a62 100644
--- a/freecad/asm3/__init__.py
+++ b/freecad/asm3/__init__.py
@@ -0,0 +1 @@
+__version__ = "0.10.2"
diff --git a/freecad/asm3/constraint.py b/freecad/asm3/constraint.py
index 3f9c6e1..7031cb0 100644
--- a/freecad/asm3/constraint.py
+++ b/freecad/asm3/constraint.py
@@ -1,7 +1,10 @@
 from collections import namedtuple
 import FreeCAD, FreeCADGui, Part
 from . import utils, gui
-from .deps import with_metaclass
+try:
+    from six import with_metaclass
+except ImportError:
+    from .deps import with_metaclass
 from .utils import objName,cstrlogger as logger, guilogger
 from .proxy import ProxyType, PropertyInfo, propGet, propGetValue
 
diff --git a/freecad/asm3/gui.py b/freecad/asm3/gui.py
index 152daf3..a1aa4db 100644
--- a/freecad/asm3/gui.py
+++ b/freecad/asm3/gui.py
@@ -1,7 +1,10 @@
 from collections import OrderedDict
 import FreeCAD, FreeCADGui
 from PySide import QtCore, QtGui
-from .deps import with_metaclass
+try:
+    from six import with_metaclass
+except ImportError:
+    from .deps import with_metaclass
 from .utils import getElementPos,objName,addIconToFCAD,guilogger as logger
 from .proxy import ProxyType
 from .FCADLogger import FCADLogger
diff --git a/freecad/asm3/sys_slvs.py b/freecad/asm3/sys_slvs.py
index a9f73fc..9dea9ea 100644
--- a/freecad/asm3/sys_slvs.py
+++ b/freecad/asm3/sys_slvs.py
@@ -1,4 +1,7 @@
-from .deps import with_metaclass
+try:
+    from six import with_metaclass
+except ImportError:
+    from .deps import with_metaclass
 from .system import System, SystemBase, SystemExtension
 from .utils import syslogger as logger, objName
 import platform, sys
diff --git a/freecad/asm3/sys_sympy.py b/freecad/asm3/sys_sympy.py
index d8b6d8b..bb82c15 100644
--- a/freecad/asm3/sys_sympy.py
+++ b/freecad/asm3/sys_sympy.py
@@ -1,6 +1,9 @@
 from collections import namedtuple
 import pprint
-from .deps import with_metaclass
+try:
+    from six import with_metaclass
+except ImportError:
+    from .deps import with_metaclass
 from .proxy import ProxyType, PropertyInfo
 from .system import System, SystemBase, SystemExtension
 from .utils import syslogger as logger, objName
diff --git a/freecad/asm3/system.py b/freecad/asm3/system.py
index 7388861..be871b7 100644
--- a/freecad/asm3/system.py
+++ b/freecad/asm3/system.py
@@ -1,6 +1,9 @@
 import os
 import FreeCAD
-from .deps import with_metaclass
+try:
+    from six import with_metaclass
+except ImportError:
+    from .deps import with_metaclass
 from .constraint import cstrName, PlaneInfo, NormalInfo
 from .utils import getIcon, syslogger as logger, objName, project2D, getNormal
 from .proxy import ProxyType, PropertyInfo
diff --git a/setup.py b/setup.py
new file mode 100644
index 0000000..fc70281
--- /dev/null
+++ b/setup.py
@@ -0,0 +1,16 @@
+from setuptools import setup
+import os
+
+version_path = os.path.join(os.path.abspath(os.path.dirname(__file__)), 
+                            "freecad", "asm3", "__init__.py")
+with open(version_path) as fp:
+    exec(fp.read())
+
+setup(name='freecad.asm3',
+      version=str(__version__),
+      packages=['freecad',
+                'freecad.asm3'],
+      url="https://github.com/realthunder/FreeCAD_assembly3",
+      description="Experimental attempt for the next generation assembly workbench for FreeCAD ",
+      install_requires=["six"],
+      include_package_data=True)
