diff --git a/setup.py b/setup.py
index d7b9687..b26fd87 100755
--- a/setup.py
+++ b/setup.py
@@ -135,11 +135,11 @@ class pivy_build(build):
                       "interfaces" + os.sep + "%s.i"
 
 
-    SOGUI = ['soqt', 'soxt', 'sogtk', 'sowin']
+    # SOGUI = ['soqt', 'soxt', 'sogtk', 'sowin']
 
     MODULES = {}
     MODULES['coin']      = ['_coin',      'coin-config',      'pivy.',        "coin"]
-    MODULES['soqt']      = ['gui._soqt',  'soqt-config',      'pivy.gui.',    "soqt"]
+    # MODULES['soqt']      = ['gui._soqt',  'soqt-config',      'pivy.gui.',    "soqt"]
 
     if sys.version_info.major < 3:
         MODULES['simvoleon'] = ['_simvoleon', 'simvoleon-config', 'pivy.',        "simvoleon"]
@@ -364,7 +364,7 @@ class pivy_build(build):
         self.check_swig_version(self.SWIG)
         self.check_coin_version()
         self.get_coin_features()
-        if self.SOGUI: self.check_gui_bindings()
+        # if self.SOGUI: self.check_gui_bindings()
         
         if 'simvoleon' in self.MODULES and self.check_simvoleon_version():
             if sys.platform == "win32":
