diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 9f1bcab..9ae7065 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -127,12 +127,12 @@ endif(NOT BUILD_TYPE)
 include_directories(${CMAKE_CURRENT_BINARY_DIR})
 
 # find BOOST and boost-python
-find_package( Boost COMPONENTS python REQUIRED)
+find_package( Boost COMPONENTS python3 REQUIRED)
 if(Boost_FOUND)
   include_directories(${Boost_INCLUDE_DIRS} ${PYTHON_INCLUDE_DIRS})
   message(STATUS "found Boost: " ${Boost_LIB_VERSION})
   message(STATUS "boost-incude dirs are: " ${Boost_INCLUDE_DIRS})
-  message(STATUS "boost-python lib is: " ${Boost_PYTHON_LIBRARY})
+  message(STATUS "boost-python lib is: " ${Boost_PYTHON3_LIBRARY})
   message(STATUS "boost_LIBRARY_DIRS is: " ${Boost_LIBRARY_DIRS})
   message(STATUS "Boost_LIBRARIES is: " ${Boost_LIBRARIES})    
 endif()
@@ -319,8 +319,8 @@ if (BUILD_PY_LIB)
     
     )
 
-  message(STATUS "linking python binary ocl.so with boost: " ${Boost_PYTHON_LIBRARY})
-  target_link_libraries(ocl ocl_common ocl_dropcutter ocl_cutters  ocl_geo ocl_algo ${Boost_LIBRARIES}  ${PYTHON_LIBRARIES} -lboost_python -lboost_system)
+  message(STATUS "linking python binary ocl.so with boost: " ${Boost_PYTHON3_LIBRARY})
+  target_link_libraries(ocl ocl_common ocl_dropcutter ocl_cutters  ocl_geo ocl_algo ${Boost_LIBRARIES}  ${PYTHON_LIBRARIES} -lboost_python3 -lboost_system)
   # 
   # this makes the lib name ocl.so and not libocl.so
   set_target_properties(ocl PROPERTIES PREFIX "") 
