diff --git a/CMakeLists.txt b/CMakeLists.txt
index 93e3f76..f52d870 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -84,7 +84,7 @@ else()
 	include/appimageupdaterdialog.hpp)
 endif()
 
-add_library(AppImageUpdaterBridge ${source})
+add_library(AppImageUpdaterBridge SHARED ${source})
 target_link_libraries(AppImageUpdaterBridge PUBLIC Qt5::Core Qt5::Network)
 target_include_directories(AppImageUpdaterBridge INTERFACE "$<BUILD_INTERFACE:. include>")
 target_include_directories(AppImageUpdaterBridge INTERFACE "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/AppImageUpdaterBridge>" )
